diff --git a/../steps/07_image_views/main.go b/../steps/08_graphics_pipeline/main.go
index 3dc703b..2f8e289 100644
--- a/../steps/07_image_views/main.go
+++ b/../steps/08_graphics_pipeline/main.go
@@ -115,7 +115,12 @@ func (app *HelloTriangleApplication) initVulkan() error {
 		return err
 	}
 
-	return app.createSwapchain()
+	err = app.createSwapchain()
+	if err != nil {
+		return err
+	}
+
+	return app.createGraphicsPipeline()
 }
 
 func (app *HelloTriangleApplication) mainLoop() error {
@@ -418,6 +423,10 @@ func (app *HelloTriangleApplication) createSwapchain() error {
 	return nil
 }
 
+func (app *HelloTriangleApplication) createGraphicsPipeline() error {
+	return nil
+}
+
 func (app *HelloTriangleApplication) chooseSwapSurfaceFormat(availableFormats []khr_surface.Format) khr_surface.Format {
 	for _, format := range availableFormats {
 		if format.Format == core1_0.FormatB8G8R8A8SRGB && format.ColorSpace == khr_surface.ColorSpaceSRGBNonlinear {
