diff --git a/../steps/25_texture_mapping/main.go b/../steps/26_depth_buffering/main.go
index 70486cf..faff3cf 100644
--- a/../steps/25_texture_mapping/main.go
+++ b/../steps/26_depth_buffering/main.go
@@ -48,7 +48,7 @@ type SwapChainSupportDetails struct {
 }
 
 type Vertex struct {
-	Position mgl32.Vec2
+	Position mgl32.Vec3
 	Color    mgl32.Vec3
 	TexCoord mgl32.Vec2
 }
@@ -95,13 +95,21 @@ func getVertexAttributeDescriptions() []core1_0.VertexInputAttributeDescription
 }
 
 var vertices = []Vertex{
-	{Position: mgl32.Vec2{-0.5, -0.5}, Color: mgl32.Vec3{1, 0, 0}, TexCoord: mgl32.Vec2{1, 0}},
-	{Position: mgl32.Vec2{0.5, -0.5}, Color: mgl32.Vec3{0, 1, 0}, TexCoord: mgl32.Vec2{0, 0}},
-	{Position: mgl32.Vec2{0.5, 0.5}, Color: mgl32.Vec3{0, 0, 1}, TexCoord: mgl32.Vec2{0, 1}},
-	{Position: mgl32.Vec2{-0.5, 0.5}, Color: mgl32.Vec3{1, 1, 1}, TexCoord: mgl32.Vec2{1, 1}},
+	{Position: mgl32.Vec3{-0.5, -0.5, 0}, Color: mgl32.Vec3{1, 0, 0}, TexCoord: mgl32.Vec2{1, 0}},
+	{Position: mgl32.Vec3{0.5, -0.5, 0}, Color: mgl32.Vec3{0, 1, 0}, TexCoord: mgl32.Vec2{0, 0}},
+	{Position: mgl32.Vec3{0.5, 0.5, 0}, Color: mgl32.Vec3{0, 0, 1}, TexCoord: mgl32.Vec2{0, 1}},
+	{Position: mgl32.Vec3{-0.5, 0.5, 0}, Color: mgl32.Vec3{1, 1, 1}, TexCoord: mgl32.Vec2{1, 1}},
+
+	{Position: mgl32.Vec3{-0.5, -0.5, -0.5}, Color: mgl32.Vec3{1, 0, 0}, TexCoord: mgl32.Vec2{0, 0}},
+	{Position: mgl32.Vec3{0.5, -0.5, -0.5}, Color: mgl32.Vec3{0, 1, 0}, TexCoord: mgl32.Vec2{1, 0}},
+	{Position: mgl32.Vec3{0.5, 0.5, -0.5}, Color: mgl32.Vec3{0, 0, 1}, TexCoord: mgl32.Vec2{1, 1}},
+	{Position: mgl32.Vec3{-0.5, 0.5, -0.5}, Color: mgl32.Vec3{1, 1, 1}, TexCoord: mgl32.Vec2{0, 1}},
 }
 
-var indices = []uint16{0, 1, 2, 2, 3, 0}
+var indices = []uint16{
+	0, 1, 2, 2, 3, 0,
+	4, 5, 6, 6, 7, 4,
+}
 
 type HelloTriangleApplication struct {
 	window *sdl.Window
@@ -154,6 +162,10 @@ type HelloTriangleApplication struct {
 	textureImageMemory core1_0.DeviceMemory
 	textureImageView   core1_0.ImageView
 	textureSampler     core1_0.Sampler
+
+	depthImage       core1_0.Image
+	depthImageMemory core1_0.DeviceMemory
+	depthImageView   core1_0.ImageView
 }
 
 func (app *HelloTriangleApplication) Run() error {
@@ -241,12 +253,17 @@ func (app *HelloTriangleApplication) initVulkan() error {
 		return err
 	}
 
-	err = app.createFramebuffers()
+	err = app.createCommandPool()
 	if err != nil {
 		return err
 	}
 
-	err = app.createCommandPool()
+	err = app.createDepthResources()
+	if err != nil {
+		return err
+	}
+
+	err = app.createFramebuffers()
 	if err != nil {
 		return err
 	}
@@ -338,6 +355,21 @@ appLoop:
 }
 
 func (app *HelloTriangleApplication) cleanupSwapChain() {
+	if app.depthImageView != nil {
+		app.depthImageView.Destroy(nil)
+		app.depthImageView = nil
+	}
+
+	if app.depthImage != nil {
+		app.depthImage.Destroy(nil)
+		app.depthImage = nil
+	}
+
+	if app.depthImageMemory != nil {
+		app.depthImageMemory.Free(nil)
+		app.depthImageMemory = nil
+	}
+
 	for _, framebuffer := range app.swapchainFramebuffers {
 		framebuffer.Destroy(nil)
 	}
@@ -499,6 +531,11 @@ func (app *HelloTriangleApplication) recreateSwapChain() error {
 		return err
 	}
 
+	err = app.createDepthResources()
+	if err != nil {
+		return err
+	}
+
 	err = app.createFramebuffers()
 	if err != nil {
 		return err
@@ -759,7 +796,7 @@ func (app *HelloTriangleApplication) createImageViews() error {
 
 	var imageViews []core1_0.ImageView
 	for _, image := range images {
-		view, err := app.createImageView(image, app.swapchainImageFormat)
+		view, err := app.createImageView(image, app.swapchainImageFormat, core1_0.ImageAspectColor)
 		if err != nil {
 			return err
 		}
@@ -772,6 +809,11 @@ func (app *HelloTriangleApplication) createImageViews() error {
 }
 
 func (app *HelloTriangleApplication) createRenderPass() error {
+	depthFormat, err := app.findDepthFormat()
+	if err != nil {
+		return err
+	}
+
 	renderPass, _, err := app.device.CreateRenderPass(nil, core1_0.RenderPassCreateInfo{
 		Attachments: []core1_0.AttachmentDescription{
 			{
@@ -784,6 +826,16 @@ func (app *HelloTriangleApplication) createRenderPass() error {
 				InitialLayout:  core1_0.ImageLayoutUndefined,
 				FinalLayout:    khr_swapchain.ImageLayoutPresentSrc,
 			},
+			{
+				Format:         depthFormat,
+				Samples:        core1_0.Samples1,
+				LoadOp:         core1_0.AttachmentLoadOpClear,
+				StoreOp:        core1_0.AttachmentStoreOpDontCare,
+				StencilLoadOp:  core1_0.AttachmentLoadOpDontCare,
+				StencilStoreOp: core1_0.AttachmentStoreOpDontCare,
+				InitialLayout:  core1_0.ImageLayoutUndefined,
+				FinalLayout:    core1_0.ImageLayoutDepthStencilAttachmentOptimal,
+			},
 		},
 		Subpasses: []core1_0.SubpassDescription{
 			{
@@ -794,6 +846,10 @@ func (app *HelloTriangleApplication) createRenderPass() error {
 						Layout:     core1_0.ImageLayoutColorAttachmentOptimal,
 					},
 				},
+				DepthStencilAttachment: &core1_0.AttachmentReference{
+					Attachment: 1,
+					Layout:     core1_0.ImageLayoutDepthStencilAttachmentOptimal,
+				},
 			},
 		},
 		SubpassDependencies: []core1_0.SubpassDependency{
@@ -801,11 +857,11 @@ func (app *HelloTriangleApplication) createRenderPass() error {
 				SrcSubpass: core1_0.SubpassExternal,
 				DstSubpass: 0,
 
-				SrcStageMask:  core1_0.PipelineStageColorAttachmentOutput,
+				SrcStageMask:  core1_0.PipelineStageColorAttachmentOutput | core1_0.PipelineStageEarlyFragmentTests,
 				SrcAccessMask: 0,
 
-				DstStageMask:  core1_0.PipelineStageColorAttachmentOutput,
-				DstAccessMask: core1_0.AccessColorAttachmentWrite,
+				DstStageMask:  core1_0.PipelineStageColorAttachmentOutput | core1_0.PipelineStageEarlyFragmentTests,
+				DstAccessMask: core1_0.AccessColorAttachmentWrite | core1_0.AccessDepthStencilAttachmentWrite,
 			},
 		},
 	})
@@ -948,6 +1004,12 @@ func (app *HelloTriangleApplication) createGraphicsPipeline() error {
 		MinSampleShading:     1.0,
 	}
 
+	depthStencil := &core1_0.PipelineDepthStencilStateCreateInfo{
+		DepthTestEnable:  true,
+		DepthWriteEnable: true,
+		DepthCompareOp:   core1_0.CompareOpLess,
+	}
+
 	colorBlend := &core1_0.PipelineColorBlendStateCreateInfo{
 		LogicOpEnabled: false,
 		LogicOp:        core1_0.LogicOpCopy,
@@ -978,6 +1040,7 @@ func (app *HelloTriangleApplication) createGraphicsPipeline() error {
 			ViewportState:      viewport,
 			RasterizationState: rasterization,
 			MultisampleState:   multisample,
+			DepthStencilState:  depthStencil,
 			ColorBlendState:    colorBlend,
 			Layout:             app.pipelineLayout,
 			RenderPass:         app.renderPass,
@@ -1000,6 +1063,7 @@ func (app *HelloTriangleApplication) createFramebuffers() error {
 			Layers:     1,
 			Attachments: []core1_0.ImageView{
 				imageView,
+				app.depthImageView,
 			},
 			Width:  app.swapchainExtent.Width,
 			Height: app.swapchainExtent.Height,
@@ -1032,6 +1096,49 @@ func (app *HelloTriangleApplication) createCommandPool() error {
 	return nil
 }
 
+func (app *HelloTriangleApplication) createDepthResources() error {
+	depthFormat, err := app.findDepthFormat()
+	if err != nil {
+		return err
+	}
+
+	app.depthImage, app.depthImageMemory, err = app.createImage(app.swapchainExtent.Width,
+		app.swapchainExtent.Height,
+		depthFormat,
+		core1_0.ImageTilingOptimal,
+		core1_0.ImageUsageDepthStencilAttachment,
+		core1_0.MemoryPropertyDeviceLocal)
+	if err != nil {
+		return err
+	}
+	app.depthImageView, err = app.createImageView(app.depthImage, depthFormat, core1_0.ImageAspectDepth)
+	return err
+}
+
+func (app *HelloTriangleApplication) findSupportedFormat(formats []core1_0.Format, tiling core1_0.ImageTiling, features core1_0.FormatFeatureFlags) (core1_0.Format, error) {
+	for _, format := range formats {
+		props := app.physicalDevice.FormatProperties(format)
+
+		if tiling == core1_0.ImageTilingLinear && (props.LinearTilingFeatures&features) == features {
+			return format, nil
+		} else if tiling == core1_0.ImageTilingOptimal && (props.OptimalTilingFeatures&features) == features {
+			return format, nil
+		}
+	}
+
+	return 0, errors.Newf("failed to find supported format for tiling %s, featureset %s", tiling, features)
+}
+
+func (app *HelloTriangleApplication) findDepthFormat() (core1_0.Format, error) {
+	return app.findSupportedFormat([]core1_0.Format{core1_0.FormatD32SignedFloat, core1_0.FormatD32SignedFloatS8UnsignedInt, core1_0.FormatD24UnsignedNormalizedS8UnsignedInt},
+		core1_0.ImageTilingOptimal,
+		core1_0.FormatFeatureDepthStencilAttachment)
+}
+
+func hasStencilComponent(format core1_0.Format) bool {
+	return format == core1_0.FormatD32SignedFloatS8UnsignedInt || format == core1_0.FormatD24UnsignedNormalizedS8UnsignedInt
+}
+
 func (app *HelloTriangleApplication) createTextureImage() error {
 	//Put image data into staging buffer
 	imageBytes, err := fileSystem.ReadFile("images/texture.png")
@@ -1094,7 +1201,7 @@ func (app *HelloTriangleApplication) createTextureImage() error {
 
 func (app *HelloTriangleApplication) createTextureImageView() error {
 	var err error
-	app.textureImageView, err = app.createImageView(app.textureImage, core1_0.FormatR8G8B8A8SRGB)
+	app.textureImageView, err = app.createImageView(app.textureImage, core1_0.FormatR8G8B8A8SRGB, core1_0.ImageAspectColor)
 	return err
 }
 
@@ -1122,13 +1229,13 @@ func (app *HelloTriangleApplication) createSampler() error {
 	return err
 }
 
-func (app *HelloTriangleApplication) createImageView(image core1_0.Image, format core1_0.Format) (core1_0.ImageView, error) {
+func (app *HelloTriangleApplication) createImageView(image core1_0.Image, format core1_0.Format, aspect core1_0.ImageAspectFlags) (core1_0.ImageView, error) {
 	imageView, _, err := app.device.CreateImageView(nil, core1_0.ImageViewCreateInfo{
 		Image:    image,
 		ViewType: core1_0.ImageViewType2D,
 		Format:   format,
 		SubresourceRange: core1_0.ImageSubresourceRange{
-			AspectMask:     core1_0.ImageAspectColor,
+			AspectMask:     aspect,
 			BaseMipLevel:   0,
 			LevelCount:     1,
 			BaseArrayLayer: 0,
@@ -1554,6 +1661,7 @@ func (app *HelloTriangleApplication) createCommandBuffers() error {
 				},
 				ClearValues: []core1_0.ClearValue{
 					core1_0.ClearValueFloat{0, 0, 0, 1},
+					core1_0.ClearValueDepthStencil{Depth: 1.0, Stencil: 0},
 				},
 			})
 		if err != nil {
@@ -1680,8 +1788,13 @@ func (app *HelloTriangleApplication) updateUniformBuffer(currentImage int) error
 	ubo.Model = mgl32.HomogRotate3D(timePeriod*mgl32.DegToRad(90.0), mgl32.Vec3{0, 0, 1})
 	ubo.View = mgl32.LookAt(2, 2, 2, 0, 0, 0, 0, 0, 1)
 	aspectRatio := float32(app.swapchainExtent.Width) / float32(app.swapchainExtent.Height)
-	ubo.Proj = mgl32.Perspective(mgl32.DegToRad(45), aspectRatio, 0.1, 10)
-	ubo.Proj[5] *= -1
+
+	near := 0.1
+	far := 10.0
+	fovy := mgl32.DegToRad(45)
+	fmn, f := far-near, float32(1./math.Tan(float64(fovy)/2.0))
+
+	ubo.Proj = mgl32.Mat4{float32(f / aspectRatio), 0, 0, 0, 0, float32(-f), 0, 0, 0, 0, float32(-far / fmn), -1, 0, 0, float32(-(far * near) / fmn), 0}
 
 	err := writeData(app.uniformBuffersMemory[currentImage], 0, &ubo)
 	return err
